### ISSUE TITLE
#include <catch2/catch.hpp> vs #include "catch.hpp"

### ISSUE THREAD

**Issue by MikeGitb** [catchorg/Catch2#1202]
Just wondering: It seems more ideomatic to me to include the library name as part of the include name instead of just using the header file name alone. On the other hand, for a single header library there is of course not a big need/difference for it unless the header name is likely to clash with a different file name.

So I was wondering whether your preference of 

    #include "catch.hpp"

over

    #include <catch2/catch.hpp>

was deliberate or more like "for historic reasons" / "didn't care".

**Comment by Quincunx271** [comment]
For single-header libraries, it's common for it to just be

```
#include <libname.hpp>
```

Alternatively, sometimes when the name is multiple words:

```
#include <super/awesome/library.hpp>
```

The code doesn't read as well, and people will be annoyed when writing it, if you had to write:

```
#include <libname/libname.hpp>
#include <super/awesome/library/library.hpp>
```

**Comment by MikeGitb** [comment]
Considering that a test file usually contains much more code than just the include I find it strange that people get annoyed if they have to type a few characters more once in a while, but I'm using an IDE with auto complete, so that might skew my view. 

Thanks for the explanation anyway. I'll consider it for my own library.

**Comment by martinmoene** [comment]
Or : #include <_namespace_/_filename_>

See also [Directories = namespaces?](https://arne-mertz.de/2016/06/organizing-directories-namespaces/#Directories_namespaces) by Arne Mertz.

**Comment by SoapGentoo** [comment]
Personally I don't mind either way from a distribution point of view - just be aware that if you're supposed to use `catch2/catch.hpp`, then trying to use it with CMake's `SYSTEM` directive will cause `-isystem /usr/include` to go up in a big explosion, due to the way GCC handles special system headers (`math.h` etc). Staying with the current approach would allow `-isystem`. Personally I don't consider `-isystem` compatibility that important, given how Catch tries to be warning-free by design.

**Comment by MikeGitb** [comment]
@SoapGentoo: Maybe I missunderstood you, but we might think of different implementation strategies:

If a packet manager / distributor wants to switch from "catch.h" to "catch2/catch.h" the way that seems most logical to me would be to move the file from "$BaseDir/catch.h" to "$BaseDir/catch2/catch.h". Any include directives should stay the same as before, so this should not pose any new problems. This might also be beneficial if someone wants to install both catch and catch2 in parallel (or it might just be more confusing).

@horenmar: Thanks for making those changes. At least for cmake I believe the default workflow for any external dependency should be

    find_package(LIB_NAME)
    target_link_libraries(my_app LIB_NAME::lib)

and then

   #include <LIB_NAME/header_file123.h>

No matter whether it is a "normal" library, a header only one or just a single header one - unfortunately standardization in that matter is exactly what c++ lacks, but it seems to be the emerging default. As you and other's pointed out, it isn't that big of a deal in the end as one can easily just copy the single headerfile in the right folder and be done with it but I prefer uniformity when possible. 

As this was mainly about satisfying my curiosity, the issue can be closed, as far as I am concerned, but if you are interested in further discussions on this you can of course leave it open.

**Comment by johnthagen** [comment]
I've been using 

```
#include "catch/catch.hpp"
```

Bumping the folder to `catch2` seems reasonable given the project has rebranded.

I've always using `< >` includes for system installed things (so if I install Boost with `apt`, I'd import as `#include <boost/...>`. When vendoring Catch2 into my own project I use `" " ` imports just like my other includes that are local to the project.

Is there a convention to use `< >` even for includes that are part of your actually source tree?

> unfortunately standardization in that matter is exactly what c++ lacks,

This.

**Comment by SoapGentoo** [comment]
@MikeGitb that's fine with me, might just be a lot of nesting for many people. @horenmar what do you say about @MikeGitb's suggestion? The only real downside is the deep hierarchy, otherwise it pleases everyone really.

**Comment by MikeGitb** [comment]
I'd be particularly interested in whether you want indeed go for `#include <catch2/catch.hpp>` or `#include 
<catch/catch.hpp>`  as I'd like to apply this to the port provided by vcpkg

**Comment by SoapGentoo** [comment]
Personally I would go with `#include <catch2/catch.hpp>`, because it allows for introducing `#include <catch3/catch.hpp>` when the API changes so drastically that you need to change significant portions of consuming code. It also makes the requirement clear at a glance without having to dig through the build system ("Oh I need modern Catch2 here").

**Comment by MikeGitb** [comment]
I'd also prefer that, but currently the CMakeLists.txt still usese `catch` which is why I held of the PR to vcpkg (which currently just puts catch.hpp directly into the include directory).

**Comment by horenmar** [comment]
We do want to go with `#include <catch2/catch.hpp>` because of the rename.

We would also prefer to switch various distribution methods at roughly the same time, so e.g. moving vcpkg to the new include path before the main repo's CMake does, would be counterproductive (as a cross-platform project could find itself needing different includes based on what distribution method it used to get Catch).

Specifically with vcpkg I already talked to Robert Schumacher offsite. The plan right now is to change its portfile to be based on the git repository, rather than downloading the header directly. This should let it easily provide `CatchConfig.cmake` and the contents from `contrib/` while staying up to date.

**Comment by MikeGitb** [comment]
That was exactly my plan, but iirc, currently the git repo / cmake based installation will install it to `include/catch` not `include` where it is currently placed in vcpkg. This is why I held off a PR to prevent two path changes in short succession.

Anyway: Good to know, you'll take care of it.

**Comment by ras0219-msft** [comment]
I've updated the vcpkg master to use catch's GitHub, however for now I've also added a forwarder from `include\catch.hpp` to `include\catch\catch.hpp` since we previously provided that as the header.

**Comment by horenmar** [comment]
I am going to keep this issue open until the change happens, so that it is easier to find.

**Comment by SoapGentoo** [comment]
@horenmar do you mind a PR to start moving the headers and installation paths?

**Comment by MikeGitb** [comment]
Any updates or an Idea when this change might happen. Is this something I can help with?

**Comment by horenmar** [comment]
@SoapGentoo @MikeGitb I was busy with non-Catch related things, but this change is now happening in `dev-new-cmake` branch.

Both pkgconfig and installed cmake package should now provide `<catch2/catch.hpp>` as path, what needs to be done is
1) Update documentation to reflect this change
2) Decide what to do when Catch is used as a subdirectory, as the current folder structure does not support this.
3+) Update external users to the new path

**Comment by MikeGitb** [comment]
Not sure why this is happening, but as far as I can tell, catch.hpp gets installed to 

    <install_prefix>/include/catch2/catch2/catch.hpp

and when using the 

    find_package(Catch2) 
    target_link_library(my_app Catch2::Catch2)`

mechanism the user has to write 

    #include <catch2/catch2/catch.hpp>

A quick fix that worked for me was to change this

https://github.com/catchorg/Catch2/blob/57ff9d054e92b36f121f4d75881d6ac9c87ff8d1/CMakeLists.txt#L73-L77
to

    target_include_directories(Catch2
      INTERFACE
        $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/single_include>
        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/catch2> 
    )

**Comment by horenmar** [comment]
@MikeGitb Interesting. I tried it locally and the install step went as supposed to.

**Comment by MikeGitb** [comment]
How did you set the installation prefix?

**Comment by horenmar** [comment]
Nevermind, d65b09fb8ae48 broke it and I tested with an older commit.

**Comment by horenmar** [comment]
@MikeGitb @SoapGentoo This is now in master, pending next release.

I also updated pkg-config support to provide the right flags to do `#include <catch2/catch.hpp>`.

**Comment by MikeGitb** [comment]
Thanks!

**Comment by philsquared** [comment]
I've only been following this with half an eye (as I've been super busy with other things, but finally getting back to some work on Catch2 for a while).
I notice that @horenmar committed 44722f9ed3d019bdff92d003f9e73333e68c503a yesterday, which seems to include most of the what's been discussed in this thread (it was a single commit which I presume was squashed from a set of branch commits?).

Anyway, what I've noticed is that a side-effect of this work is that most of the SelfTest project related stuff in CMakeLists.txt has been moved into a CMakeLists.txt in the `projects` directory. This means that the `SelfTest` executable is built to a `projects` subdirectory in the CMake output dirs (at least on my Mac). Currently this breaks the approval testing, which is still looking for them in the CMake output root - it seems to be "working" because the previously built binaries happen to be there (I'm guessing Travis has been reusing build directories?) - or perhaps I'm missing something else?

Anyway - the approval test scripts can be changed, but I wanted to see if this is the direction we want to go in (and what the reasoning is)? I'm guessing the motivation is to separate out the parts of the CMake build that relate to self-testing from those related to install? If so is this really the only/ best way to do it (I'm fine if it is - just want to be sure that's been considered). And if so I think I'd prefer that sub-directory to be something more like `SelfTest` than `projects`.

**Comment by philsquared** [comment]
(in fact I can see that approval tests are working on Travis et al because that path is being supplied explicitly from CMakeLists.txt - so less of an issue there - it's my local runs that suffer - but the main points stand).

**Comment by horenmar** [comment]
The CMake files have been getting increasingly large, making them annoying to work with., so they've been split into pieces.

**Comment by philsquared** [comment]
which is fine - except that it does change the build output

**Comment by MikeGitb** [comment]
FYI: I added a PR to vcpkg to update catch to release 2.3: https://github.com/Microsoft/vcpkg/pull/3959

**Comment by MikeGitb** [comment]
So, what needs to be done/decided to close this issue?

**Comment by MikeGitb** [comment]
ping @philsquared

Imho the split up of the cmake file and other directory names have little to do with the official include path. So if the are things that need discussion that should probably happen in a separate issue.

**Comment by horenmar** [comment]
Yeah, this is done.
